tests/integration/relations/opensearch_provider/application-charm/src/charm.py

```diff
@@ -0,0 +1,239 @@
+        """Event triggered when a database was created for this application."""
+        logging.info(f"first database credentials: {event.username} {event.password}")
+
+    def _on_first_database_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"first database endpoints have been changed to: {event.endpoints}")
+
+    # Second database events observers.
+    def _on_second_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        logger.info(f"second database credentials: {event.username} {event.password}")
+
+    def _on_second_database_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"second database endpoints have been changed to: {event.endpoints}")
+
+    # ==============
+    #  Action hooks
+    # ==============
+    def _on_run_request_action(self, event: ActionEvent):
+        logger.info(event.params)
+        relation = self.first_database
+        relation_id = event.params["relation-id"]
+        databag = relation.fetch_relation_data()[relation_id]
+        method = event.params["method"]
+        endpoint = event.params["endpoint"]
+        payload = event.params.get("payload", None)
+        if payload:
+            payload = payload.replace("\\", "")
+
+        username = databag.get("username")
+        password = databag.get("password")
+        host = databag.get("endpoints").split(",")[0]
+        host_addr = host.split(":")[0]
+        port = host.split(":")[1]
+
+        logger.info(f"sending {method} request to {endpoint}")
+        try:
+            response = self.request(method, endpoint, port, username, password, host_addr, payload)
+        except OpenSearchHttpError as e:
+            response = [str(e)]
+        logger.info(response)
+
+        event.set_results({"results": json.dumps(response)})
+
+    # =================================
+    #  Opensearch connection functions
+    # =================================
+
+    def relation_request(
+        self,
+        relation_id: int,
+        method: str,
+        endpoint: str,
+        payload: Optional[Dict[str, any]] = None,
+    ) -> Union[Dict[str, any], List[any]]:
+        """Make an HTTP request to a specific relation."""
+        databag = self.first_database.fetch_relation_data()[relation_id]
+        logging.error(databag)
+        username = databag.get("username")
+        password = databag.get("password")
+        endpoints = databag.get("endpoints", "").split(",")
+
+        if None in [username, password] or len(endpoints) == 0:
+            raise OpenSearchHttpError
+
+        host = endpoints[0].split(":")[0]
+        port = int(endpoints[0].split(":")[1])
```

Comment:
nit: if you want it shorter `host, port = endpoints[0].split(":")` - but then you need to cast the port to int before passing it to `request`